include/CL/sycl/detail/small_array.hpp

```diff
@@ -74,10 +85,10 @@ template <typename BasicType,
 struct small_array : std::array<BasicType, Dims>,
   // To have all the usual arithmetic operations on this type
   boost::euclidean_ring_operators<FinalType>,
+  boost::euclidean_ring_operators<FinalType, size_t>,
```

Comment:
It is not clear what you are implementing here.
Adding a broadcast mechanism from 1 element to all the element vectors ?
I tried instead of this to remove the `explicit` in https://github.com/triSYCL/triSYCL/blob/dbc465271f3f8f4a39385a055f430658818ea52e/include/CL/sycl/detail/small_array.hpp#L349 and https://github.com/triSYCL/triSYCL/blob/dbc465271f3f8f4a39385a055f430658818ea52e/include/CL/sycl/detail/small_array.hpp#L322 and it looks it breaks a lot of things... Curious.

Reply:
The spec has operator overload for + , - , * , / , % for both 

id<dimensions> operatorOP(const id<dimensions> &rhs) const;
id<dimensions> operatorOP(const size_t &rhs) const;

So I'm implemented the latter using the boost euclidean ring operators.

I don't you can remove the explicit there because there is no mention in the spec of that type of constructor and having it implicity start happening is bound to start generating a bunch of templating ambiguity problems (story of my C++ life).

Reply:
Those non-explicitly cause ambiguous conversions around parallel_for between id and ranges, which makes sense if you give it a id<1> and it doesn't know whether to pick range<1>, range<2> or range<3>

Reply:
I've changed this locally to BasicType, and I tried removing it and the test_id fails to build due to missing functions.